src/features/device-state/routes/state-patch.ts

```diff
@@ -256,6 +261,57 @@ const shouldMetricsUpdate = (() => {
 	};
 })();
 
+type ImageInstallUpdateBody = {
+	status: string;
+	is_provided_by__release: number;
+	download_progress?: number;
+};
+const shouldUpdateImageInstall = (() => {
+	const lastImageInstallReport = createMultiLevelStore<
+		ImageInstallUpdateBody & { updateTime: number }
+	>(
+		'lastImageInstallReport',
+		{
+			ttl: DOWNLOAD_PROGRESS_MAX_REPORT_INTERVAL_SECONDS,
+		},
+		false,
+	);
+	const DOWNLOAD_PROGRESS_MAX_REPORT_INTERVAL =
+		DOWNLOAD_PROGRESS_MAX_REPORT_INTERVAL_SECONDS * 1000;
+	return async (imageId: number, body: ImageInstallUpdateBody) => {
+		const key = `${imageId}`;
+		const lastReport = await lastImageInstallReport.get(key);
+		const now = Date.now();
+		if (
+			lastReport == null ||
+			// If the entry has expired then it means we should actually do the report
+			lastReport.updateTime + DOWNLOAD_PROGRESS_MAX_REPORT_INTERVAL < now ||
+			// Or if the status has changed
+			lastReport.status !== body.status ||
+			// Or if the release has changed
+			lastReport.is_provided_by__release !== body.is_provided_by__release ||
+			// Or if the download progress has hit a milestone...
+			// From not downloading to downloading
+			(lastReport.download_progress == null &&
+				body.download_progress != null) ||
+			// From downloading to not downloading
+			(lastReport.download_progress != null &&
+				body.download_progress == null) ||
+			// Hits 100%
+			body.download_progress === 100
+		) {
+			// And we add a new entry
+			await lastImageInstallReport.set(key, {
+				download_progress: lastReport?.download_progress,
```

Comment:
Why are we only assigning the `download_progress`?
Could/Shouldn't this be just destructuring the whole `lastReport`?

Reply:
Right I meant to add a comment to this as well, but basically we want to keep the last reported download progress if the current report doesn't include it

Reply:
https://github.com/balena-io/open-balena-api/pull/888

Reply:
@Page- my question was mostly why we care only about the `download_progress` and we didn't do:
```
				...lastReport,
				...body,
				updateTime: now,
```